File: Week_08/G20190343020039/LeetCode_63_0039.py
from typing import List

class Solution:
  def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
    rows = len(obstacleGrid)
    cols = len(obstacleGrid[0])
    dp = [[0 for _ in range(cols)] for _ in range(rows)]

    if obstacleGrid[rows-1][cols-1] == 0:
      dp[rows-1][cols-1] = 1

    for row in range(rows-2, -1, -1):
      if obstacleGrid[row][cols-1] == 0 and dp[row+1][cols-1] == 1:
        dp[row][cols-1] = 1

    for col in range(cols-2, -1, -1):
      if obstacleGrid[rows-1][col] == 0 and dp[rows-1][col+1] == 1:
        dp[rows-1][col] = 1

    for row in range(rows-2, -1, -1):
      for col in range(cols-2, -1, -1):
        if obstacleGrid[row][col] == 1:
          dp[row][col] = 0
        else:
          dp[row][col] = dp[row+1][col] + dp[row][col+1]

    return dp[0][0]


s = Solution()
result = s.uniquePathsWithObstacles([[0,0,0],[0,1,0],[0,0,0]])
print(result)

File: Week_05/G20190343020039/LeetCode_63_0039.js
/**
 * @param {number[][]} obstacleGrid
 * @return {number}
 */
var uniquePathsWithObstacles = function(obstacleGrid) {
  let rows = obstacleGrid.length;
  let cols = obstacleGrid[0].length;
  let dp = new Array(rows).fill(null).map(() => {
    return new Array(cols).fill(0);
  });

  dp[rows-1][cols-1] = obstacleGrid[rows-1][cols-1] === 0 ? 1 : 0;

  for (let row = rows - 2; row >= 0; row -= 1) {
    dp[row][cols-1] = (obstacleGrid[row][cols-1] === 0
      && dp[row+1][cols-1] === 1) ? 1 : 0;
  }

  for (let col = cols - 2; col >= 0; col -= 1) {
    dp[rows-1][col] = (obstacleGrid[rows-1][col] === 0
      && dp[rows-1][col+1] === 1) ? 1 : 0;
  }

  for (let row = rows - 2; row >= 0; row -= 1) {
    for (let col = cols - 2; col >= 0; col -= 1) {
      if (obstacleGrid[row][col] === 1) {
        dp[row][col] = 0;
      } else {
        dp[row][col] = dp[row+1][col] + dp[row][col+1];
      }
    }
  }

  return dp[0][0];
};
